draco/analysis/delayopt.py

```diff
@@ -0,0 +1,593 @@
+        fsel = np.arange(Nf)
+    elif len(fsel) != Nf:
+        raise ValueError(
+            "Length of frequency selection must match frequencies passed. "
+            f"{len(fsel)} != {data.shape[-1]}"
+        )
+
+    # Construct the Fourier matrix
+    F = fourier_matrix(N, fsel)
+
+    # Compute the covariance matrix of the data
+    # Window the frequency data if requested
+    if window is not None:
+        # Construct the window function
+        x = fsel * 1.0 / N
+        w = tools.window_generalised(x, window=window)
+
+        # Apply to the projection matrix and the data
+        F *= w[:, np.newaxis]
+        data = data * w[np.newaxis, :]
+
+    X = data.T @ data.conj()
+    X /= nsamp
+
+    # Construct the noise matrix from the diagonal of its inverse
+    Nm = np.diag(tools.invert_no_zero(Ni))
+
+    # Mask out any completely missing frequencies
+    F[Ni == 0] = 0.0
+
+    # Use the pseudo-inverse to give a starting point for the optimiser
+    if initial_S is None:
+        lsi = np.log((data @ la.pinv(F.T, rtol=1e-3)).var(axis=0))
+    else:
+        lsi = np.log(initial_S)
+
+    optfunc = AddFunctions(
+        [
+            LogLikePS(X, F, Nm, nsamp, exact_hessian=True),
+            GaussianProcessPrior(
+                N, alpha=5, width=3.0, kernel="gaussian", a=5.0, reg=1e-8
+            ),
+        ]
+    )
+
+    samples = []
+
+    # This callback is for getting the intermediate samples such that we can access
+    # convergence of the solution
+    def _get_intermediate(xk):
+        samples.append(np.exp(xk))
+
+    success = False
+    try:
+        res = minimize(
+            optfunc.value,
+            x0=lsi,
+            jac=optfunc.gradient,
+            hess=optfunc.hessian,
+            method="Newton-CG",
+            options={"maxiter": maxiter, "xtol": tol},
+            callback=_get_intermediate,
+        )
+        success = res.success
+
+    # LinAlgError gets thrown for certain baselines in _precompute during a Cholesky decomposition
+    # of the covariance matrix (used in likelihood computation) when the covariance matrix isn't
+    # positive definite. This appears to happen after one of the minimization parameters blows up,
+    # likely from a numerical instability somewhere in the gradient/hessian computation. This has
+    # only been observed to affect a small number of (almost entirely masked) baselines that have
+    # very low weights.
+    # ValueError also occasionally gets thrown from (almost certainly) the same numerical
+    # instability: one of the minimization parameters blows up and causes an overflow when taking
+    # np.exp() of the minimization parameter to get the delay spectrum. The ValueError gets thrown
+    # again in the Cholesky decomposition when numpy runs a check_finite on the covariance (which
+    # at that point contains infs/nans).
+    except (la.LinAlgError, ValueError):
+        pass  # success already set to false if this executes.
```

Comment:
While this is true and works as expected, I think it makes the code a bit clearer to just explicitly set `success = False` here rather than just having the comment